### PR TITLE
rds_instance: Add purge_security_groups

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

rds_instance: Add `purge_security_groups` parameter for both `db_security_groups` and `vpc_security_groups_ids`.
Should fix: https://github.com/ansible-collections/community.aws/issues/385

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

rds_instance

##### ADDITIONAL INFORMATION

According to the boto3 documentation https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html#RDS.Client.describe_db_instances,  `describe_db_instances()` provides  this information for the associated security groups
```
'DBSecurityGroups': [
                {
                    'DBSecurityGroupName': 'string',
                    'Status': 'string'
                },
            ],
 'VpcSecurityGroups': [
                {
                    'VpcSecurityGroupId': 'string',
                    'Status': 'string'
                },
            ],
```
With this patch the status is reflected accordingly as showed in the following example.
```
- name: Create a DB instance in the VPC with two security groups
   rds_instance:
     id: "{{ instance_id }}"
     state: present
     engine: mariadb
     username: "{{ username }}"
     password: "{{ password }}"
     db_instance_class: "{{ db_instance_class }}"
     allocated_storage: "{{ allocated_storage }}"
     vpc_security_group_ids:
            - "sg-1"
            - "sg-2"

```
This gives:
```
....
"vpc_security_groups": [
        {
            "status": "active",
            "vpc_security_group_id": "sg-1"
        },
        {
            "status": "active",
            "vpc_security_group_id": "sg-2"
        }
    ]
```
And then if we want to set the a new security group `sg-3` (purge_security_groups=True by default).
```
- name: Add a new security group (purge_security_groups=True by default)
   rds_instance:
     id: "{{ instance_id }}"
     state: present
     vpc_security_group_ids:
            - "sg-3"
```

The log produced:
```
"vpc_security_groups": [
        {
            "status": "removing",
            "vpc_security_group_id": "sg-1"
        },
        {
            "status": "removing",
            "vpc_security_group_id": "sg-2"
        },
        {
            "status": "adding",
            "vpc_security_group_id": "sg-3"
        }
    ]
```

Since some vpc security groups are reported "removing" status, do we want to show them? Others are also with "adding" state. I was wondering if there is a way to only show security groups with an active status or however wait until a security group passes from "adding" to "active".

### CODE REVIEW

**Comment by alinabuzachis** [comment]
@s-hertel Could you please take a look at this proposal?

**Comment by s-hertel** [comment]
@alinabuzachis Yes, will review in a day or two. Thanks for fixing this!

File: plugins/modules/rds_instance.py [hunk 1]
```diff
@@ -938,7 +956,9 @@ def get_changing_options_with_inconsistent_keys(modify_params, instance, purge_c
         # TODO: allow other purge_option module parameters rather than just checking for things to add
         if isinstance(current_option, list):
             if isinstance(desired_option, list):
-                if set(desired_option) <= set(current_option):
+                if set(desired_option) < set(current_option):
```
Comment (s-hertel) [inline plugins/modules/rds_instance.py:959]: The `continue` below used to happen for `<=`. You could maybe consolidate the special case into a single conditional and add an elif for the generic case.

```python
if (
    set(dedirect_option) < set(current_option) and
    option in ('DBSecurityGroups', 'VpcSecurityGroupIds',) and
    purge_security_groups
):
    ...
elif set(desired_option) <= set(current_option):
    ...

**Review by s-hertel — COMMENTED** [review]
Looks great!

> I was wondering if there is a way to only show security groups with an active status or however wait until a security group passes from "adding" to "active".

You could add a waiter to return only 'active' security groups by using a pathAll matcher - something like
```python
"DBInstanceActiveSecurityGroups": {
    "delay": 20,
    "maxAttempts": 60,
    "operation": "DescribeDBInstances",
    "acceptors": [
        {
            "state": "success",
            "matcher": "pathAll",
            "argument": "DBInstances[].VpcSecurityGroups[].Status",
            "expected": "active"  # check if this is lowercase pre-camel_to_snake_case
        },
        {
            "state": "success",
            "matcher": "pathAll",
            "argument": "DBInstances[].DBSecurityGroups[].Status",
            "expected": "active"
        },
}
```
But I actually like this as it is, so it's your call/others can weigh in if that would be useful.

File: plugins/modules/rds_instance.py (hunk 1 above)
Comment (alinabuzachis) [reply plugins/modules/rds_instance.py]: @ s-hertel Thank you so much for reviewing it. In fact, I wasn't sure whether to return all security groups or only those with added/active status. Thanks for explaining and it makes sense that the others have the freedom to decide. I applied this change for the conditional statement.

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml [hunk 2]
```diff
@@ -73,18 +73,70 @@
...
+          apply_immediately: true
+          purge_security_groups: false
+        register: result
+
+      - assert:
+          that:
+            - result.changed
+            - "result.db_instance_identifier == '{{ instance_id }}'"
+            - "result.vpc_security_groups | selectattr('status', 'in', ['active', 'adding']) | list | length == 3"
+
```
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:109]:
```suggestion

```

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml [hunk 3]
```diff
@@ -73,18 +73,70 @@
...
+          apply_immediately: true
+          purge_security_groups: false
+        check_mode: true
         register: result
 
       - assert:
           that:
             - result.changed
+            - "result.db_instance_identifier == '{{ instance_id }}'"
+
```
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:93]:
```suggestion

```

**Review by tremble — CHANGES_REQUESTED** [review]
Generally looks good.  A couple of white-space nit-picks.

I'm currently running the tests locally, assuming they pass I'll get this merged.

**Review by tremble — APPROVED** [review]

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml [hunk 4]
```diff
@@ -73,18 +73,68 @@
...
             - "{{ sgs_result.results.2.group_id }}"
+          apply_immediately: true
+          purge_security_groups: false
+        check_mode: true
         register: result
 
       - assert:
           that:
             - result.changed
+            - "result.db_instance_identifier == '{{ instance_id }}'"
```
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:92]:
```suggestion
            - "result.db_instance_identifier == '{{ instance_id }}'"

```

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml [hunk 5]
```diff
@@ -73,18 +73,68 @@
...
+            - "{{ sgs_result.results.2.group_id }}"
+          apply_immediately: true
+          purge_security_groups: false
+        register: result
+
+      - assert:
+          that:
+            - result.changed
+            - "result.db_instance_identifier == '{{ instance_id }}'"
+            - "result.vpc_security_groups | selectattr('status', 'in', ['active', 'adding']) | list | length == 3"
```
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:107]:
```suggestion
            - "result.vpc_security_groups | selectattr('status', 'in', ['active', 'adding']) | list | length == 3"

```

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml [hunk 6]
```diff
@@ -73,18 +73,68 @@
...
+            - "{{ sgs_result.results.2.group_id }}"
+          apply_immediately: true
+          purge_security_groups: false
+        register: result
+
+      - assert:
+          that:
+            - not result.changed
+            - "result.db_instance_identifier == '{{ instance_id }}'"
+
```
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:122]:
```suggestion

```

**Review by tremble — APPROVED** [review]

File: tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml (hunk 6 above)
Comment (tremble) [inline tests/integration/targets/rds_instance/tasks/test_vpc_security_groups.yml:122]:
```suggestion
            - "result.db_instance_identifier == '{{ instance_id }}'"

```

**Comment by tremble** [comment]
Tests passed successfully after unrelated fixes.

**Pull request merged**